docs/articles/extending/multihost-utilities.rst

```diff
@@ -0,0 +1,369 @@
+            """
+            if self._users:
+                cmd = "\n".join([f"userdel '{x}' --force --remove" for x in self._users]) + "\n"
+                self.host.conn.run("set -e\n\n" + cmd)
+
+            self._users = self._states.pop()
+
+        def add_local_user(
+            self,
+            *,
+            name: str,
+            uid: int | None = None,
+            gid: int | None = None,
+            password: str | None = "Secret123",
+            home: str | None = None,
+            gecos: str | None = None,
+            shell: str | None = None,
+        ) -> Self:
+            """
+            Create new local user.
+
+            :param uid: User id, defaults to None
+            :type uid: int | None, optional
+            :param gid: Primary group id, defaults to None
+            :type gid: int | None, optional
+            :param password: Password, defaults to 'Secret123'
+            :type password: str, optional
+            :param home: Home directory, defaults to None
+            :type home: str | None, optional
+            :param gecos: GECOS, defaults to None
+            :type gecos: str | None, optional
+            :param shell: Login shell, defaults to None
+            :type shell: str | None, optional
+            :return: Self.
+            :rtype: Self
+            """
+            if home is not None:
+                self.fs.backup(home)
+
+            args: CLIBuilderArgs = {
+                "name": (self.cli.option.POSITIONAL, name),
+                "uid": (self.cli.option.VALUE, uid),
+                "gid": (self.cli.option.VALUE, gid),
+                "home": (self.cli.option.VALUE, home),
+                "gecos": (self.cli.option.VALUE, gecos),
+                "shell": (self.cli.option.VALUE, shell),
+            }
+
+            passwd = f" && passwd --stdin '{name}'" if password else ""
+            self.logger.info(f'Creating local user "{name}" on {self.host.hostname}')
+            self.host.conn.run(self.cli.command("useradd", args) + passwd, input=password, log_level=ProcessLogLevel.Error)
+
+            self._users.append(name)
+
+            return self
+
+Creating more setup-scopes in tests
+-----------------------------------
+
+It is possible to enter the reentrant utilities multiple times in tests as well
+using the ``with`` statement.
+
+.. code-block:: python
+    :caption: Reentrant utility in tests
+    :emphasize-lines: 3,5,7
+    :linenos:
+
+        @pytest.mark.topology(...)
+        def test_ad_hoc_util(example: ExampleRole) -> None:
+            with example.fs as fs_a:
+                fs_a.write("/root/test", "content_a")
+                with fs_a as fs_b:
+                    fs_b.write("/root/test", "content_b")
+                    with fs_b as fs_c:
+                        fs_c.write("/root/test", "content_c")
+                    assert fs_b.read("/root/test") == "content_b"
+                assert fs_a.read("/root/test") == "content_a"
```

Comment:
I will be honest, I don't understand neither the explanation nor the code. Why would I need several setup scopes? What does the code try to solve?

Reply:
You don't understand this section or the whole purpose of MultihostReentrantUtility?

Reply:
I think the first question to answer is: what is the use case where multiple setup scopes are needed?

Reply:
The main purpose and the reason why it was implemented was to allow to use MultihostUtility (especially the fs and svc) in hosts. So if you do use fs in MultihostHost.pytest_setup it is reverted after MultihostHost.pytest_teardown. If you us fs in MultihostHost.setup it is reverted after MultihostHost.teardown. Similar for topologies.

```
pytest_setup
    setup
    teardown
pytest_teardown
```

The fact that you can create more scope inside test is for free. I don't have any real use case for it though.

Reply:
mmm ok, so everything is theoretical

Reply:
theoretical is only to use in in tests (i.e. this particular section). Otherwise we already use it inside sssd framework. But I will rewrite this and focus more on the real use case.

Reply:
The current version makes more sense